ItemDefinitionGroup does not support attribute metadata

### Steps to reproduce

Either include a project sample, attach a zipped project, or provide IDE / CLI steps to create the project and repro the behaviour. Example of a project sample:

Project file
```xml
<Project>
  <ItemDefinitionGroup>
    <Foo Bar="Baz" />
  </ItemDefinitionGroup>
  <Target Name="Build" />
</Project>
```

### Expected  behavior
The project builds successfully.

### Actual behavior
The project does not build successfully.

```
Build FAILED.

"E:\tmp\test.proj" (default target) (1) ->
  E:\tmp\test.proj(3,10): error MSB4066: The attribute "Bar" in element <Foo> is unrecognized.
```

Note that the following *does* work as expected, but the above should be equivalent to this:
```xml
<Project>
  <ItemDefinitionGroup>
    <Foo>
      <Bar>Baz</Bar>
    </Foo>
  </ItemDefinitionGroup>
  <Target Name="Build" />
</Project>
```

Attribute metadata works in `<ItemGroup>` but not `<ItemDefinitionGroup>` it seems.
### Environment data
`msbuild /version` output: 15.9.14-preview+gae560092f5

OS info: Windows 10

--- a/ref/Microsoft.Build/net/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/net/Microsoft.Build.cs
@@ -123,7 +123,9 @@ namespace Microsoft.Build.Construction
         public string ItemType { get { throw null; } }
         public System.Collections.Generic.ICollection<Microsoft.Build.Construction.ProjectMetadataElement> Metadata { get { throw null; } }
         public Microsoft.Build.Construction.ProjectMetadataElement AddMetadata(string name, string unevaluatedValue) { throw null; }
+        public Microsoft.Build.Construction.ProjectMetadataElement AddMetadata(string name, string unevaluatedValue, bool expressAsAttribute) { throw null; }
         protected override Microsoft.Build.Construction.ProjectElement CreateNewInstance(Microsoft.Build.Construction.ProjectRootElement owner) { throw null; }
+        protected override bool ShouldCloneXmlAttribute(System.Xml.XmlAttribute attribute) { throw null; }
     }
     public partial class ProjectItemDefinitionGroupElement : Microsoft.Build.Construction.ProjectElementContainer
     {

--- a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
@@ -123,7 +123,9 @@ namespace Microsoft.Build.Construction
         public string ItemType { get { throw null; } }
         public System.Collections.Generic.ICollection<Microsoft.Build.Construction.ProjectMetadataElement> Metadata { get { throw null; } }
         public Microsoft.Build.Construction.ProjectMetadataElement AddMetadata(string name, string unevaluatedValue) { throw null; }
+        public Microsoft.Build.Construction.ProjectMetadataElement AddMetadata(string name, string unevaluatedValue, bool expressAsAttribute) { throw null; }
         protected override Microsoft.Build.Construction.ProjectElement CreateNewInstance(Microsoft.Build.Construction.ProjectRootElement owner) { throw null; }
+        protected override bool ShouldCloneXmlAttribute(System.Xml.XmlAttribute attribute) { throw null; }
     }
     public partial class ProjectItemDefinitionGroupElement : Microsoft.Build.Construction.ProjectElementContainer
     {

--- a/src/Build.OM.UnitTests/Construction/ProjectItemDefinitionElement_Tests.cs
+++ b/src/Build.OM.UnitTests/Construction/ProjectItemDefinitionElement_Tests.cs
@@ -9,6 +9,7 @@ using System.Text;
 using System.Xml;
 
 using Microsoft.Build.Construction;
+using Microsoft.Build.Exceptions;
 using Microsoft.Build.Shared;
 using Xunit;
 
@@ -96,8 +97,7 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             string content = @"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
                         <ItemDefinitionGroup>
-                            <i1>
-                                <m1>v1</m1>
+                            <i1 m1='v1'>
                                 <m2 Condition='c'>v2</m2>
                                 <m1>v3</m1>
                             </i1>
@@ -119,6 +119,82 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             Assert.Equal("c", metadata[1].Condition);
             Assert.Equal("m1", metadata[2].Name);
             Assert.Equal("v3", metadata[2].Value);
+        }
+
+        /// <summary>
+        /// Reads metadata as attributes that wouldn't be
+        /// metadata on items
+        /// </summary>
+        [Theory]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i Include='inc' />
+                        </ItemDefinitionGroup> 
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i Update='upd' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i Remove='rem' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i Exclude='excl' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i KeepMetadata='true' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i RemoveMetadata='true' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i KeepDuplicates='true' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i cOndiTion='true' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i LabeL='text' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        public void DoNotReadInvalidMetadataAttributesOrAttributesValidOnItems(string content)
+        {
+            Assert.Throws<InvalidProjectFileException>(() =>
+            {
+                ProjectRootElement.Create(XmlReader.Create(new StringReader(content)));
+            });
         }
 
         /// <summary>

--- a/src/Build.OM.UnitTests/Construction/ProjectMetadataElement_Tests.cs
+++ b/src/Build.OM.UnitTests/Construction/ProjectMetadataElement_Tests.cs
@@ -84,17 +84,23 @@ namespace Microsoft.Build.UnitTests.OM.Construction
            );
         }
 
-        [Fact]
-        public void ReadInvalidNameAsAttribute()
-        {
-            string content = @"
+        [Theory]
+        [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
                         <ItemGroup>
                             <i Include='i1' " + "\u03A3" + @"='v1' />
                         </ItemGroup>
                     </Project>
-                ";
-
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i " + "\u03A3" + @"='v1' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        public void ReadInvalidNameAsAttribute(string content)
+        {
             Assert.Throws<InvalidProjectFileException>(() =>
             {
                 ProjectRootElement.Create(XmlReader.Create(new StringReader(content)));
@@ -125,19 +131,25 @@ namespace Microsoft.Build.UnitTests.OM.Construction
            );
         }
 
-        [Fact]
-        public void ReadInvalidBuiltInNameAsAttribute()
-        {
-            Assert.Throws<InvalidProjectFileException>(() =>
-            {
-                string content = @"
+        [Theory]
+        [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
                         <ItemGroup>
                             <i Include='i1' Filename='v1'/>
                         </ItemGroup>
                     </Project>
-                ";
-
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i Filename='v1'/>
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        public void ReadInvalidBuiltInNameAsAttribute(string content)
+        {
+            Assert.Throws<InvalidProjectFileException>(() =>
+            {
                 ProjectRootElement.Create(XmlReader.Create(new StringReader(content)));
             }
            );
@@ -169,19 +181,25 @@ namespace Microsoft.Build.UnitTests.OM.Construction
         /// <summary>
         /// Read metadatum with invalid built-in element name
         /// </summary>
-        [Fact]
-        public void ReadInvalidBuiltInElementNameAsAttribute()
-        {
-            Assert.Throws<InvalidProjectFileException>(() =>
-            {
-                string content = @"
+        [Theory]
+        [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
                         <ItemGroup>
                             <i Include='i1' PropertyGroup='v1' />
                         </ItemGroup>
                     </Project>
-                ";
-
+                ")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i PropertyGroup='v1' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ")]
+        public void ReadInvalidBuiltInElementNameAsAttribute(string content)
+        {
+            Assert.Throws<InvalidProjectFileException>(() =>
+            {
                 ProjectRootElement.Create(XmlReader.Create(new StringReader(content)));
             }
            );
@@ -248,7 +266,7 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             metadatum.ExpressedAsAttribute = true;
 
             Assert.Throws<InvalidProjectFileException>(() =>
-            {                
+            {
                 metadatum.Name = "Include";
             }
            );
@@ -299,6 +317,32 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             Assert.Throws<InvalidProjectFileException>(() =>
             {
                 item.AddMetadata("Include", "v1", true);
+            });
+        }
+
+        [Fact]
+        public void AddMetadataAsAttributeToItemDefinitionIllegalName()
+        {
+            string project = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i1/>
+                        </ItemDefinitionGroup>
+                    </Project>
+                ";
+
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(project)));
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+
+            var itemDefinition = itemDefinitions.First();
+
+            Assert.Throws<InvalidProjectFileException>(() =>
+            {
+                itemDefinition.AddMetadata("Include", "v1", true);
             });
         }
 
@@ -403,6 +447,30 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             Assert.True(metadata.ExpressedAsAttribute);
         }
 
+        [Fact]
+        public void ReadMetadataAsAttributeOnItemDefinition()
+        {
+            string project = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i1 m1='v1' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(project)));
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+        }
+
         [Theory]
         [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
@@ -431,6 +499,30 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             Assert.Equal(1, items[0].Metadata.Count);
 
             var metadata = items[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal(@"<&>""", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+        }
+
+        [Fact]
+        public void ReadMetadataAsAttributeOnItemDefinitionWithSpecialCharacters()
+        {
+            var project = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                        <ItemDefinitionGroup>
+                            <i1 m1='&lt;&amp;>""' />
+                        </ItemDefinitionGroup>
+                    </Project>
+                ";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(project)));
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
             Assert.Equal("m1", metadata.Name);
             Assert.Equal(@"<&>""", metadata.Value);
             Assert.True(metadata.ExpressedAsAttribute);
@@ -495,6 +587,53 @@ namespace Microsoft.Build.UnitTests.OM.Construction
 
             string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
                 ObjectModelHelpers.CleanupFileContents(updatedProject);
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
+        [Fact]
+        public void UpdateMetadataValueAsAttributeOnItemDefinition()
+        {
+            var projectContents = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1` />
+                        </ItemDefinitionGroup>
+                    </Project>";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(
+                    new StringReader(ObjectModelHelpers.CleanupFileContents(projectContents))),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            metadata.Value = "v2";
+
+            Assert.True(project.IsDirty);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                              ObjectModelHelpers.CleanupFileContents(@"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v2` />
+                        </ItemDefinitionGroup>
+                    </Project>");
             string actual = writer.ToString();
 
             VerifyAssertLineByLine(expected, actual);
@@ -568,6 +707,53 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             VerifyAssertLineByLine(expected, actual);
         }
 
+        [Fact]
+        public void UpdateMetadataValueAsAttributeOnItemDefinitionWithSpecialCharacters()
+        {
+            var projectContents = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1` />
+                        </ItemDefinitionGroup>
+                    </Project>";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(
+                    new StringReader(ObjectModelHelpers.CleanupFileContents(projectContents))),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            metadata.Value = @"<&>""";
+
+            Assert.True(project.IsDirty);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                              ObjectModelHelpers.CleanupFileContents(@"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`&lt;&amp;&gt;&quot;` />
+                        </ItemDefinitionGroup>
+                    </Project>");
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
         [Theory]
         [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
@@ -619,7 +805,7 @@ namespace Microsoft.Build.UnitTests.OM.Construction
                 ProjectCollection.GlobalProjectCollection,
                 preserveFormatting: true);
             ProjectItemGroupElement itemGroup = (ProjectItemGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemGroupElement);
-            
+
             var project = new Project(projectElement);
 
             var items = Helpers.MakeList(itemGroup.Items);
@@ -628,6 +814,67 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             Assert.Equal(1, items[0].Metadata.Count);
 
             var metadata = items[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.False(metadata.ExpressedAsAttribute);
+
+            metadata.ExpressedAsAttribute = true;
+
+            Assert.True(project.IsDirty);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                ObjectModelHelpers.CleanupFileContents(updatedProject);
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
+        [Theory]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1>
+                              <m1>v1</m1>
+                            </i1>
+                        </ItemDefinitionGroup>
+                    </Project>",
+        @"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1` />
+                        </ItemDefinitionGroup>
+                    </Project>")]
+        [InlineData(@"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1><m1>v1</m1></i1>
+                        </ItemDefinitionGroup>
+                    </Project>",
+        @"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1` />
+                        </ItemDefinitionGroup>
+                    </Project>")]
+        public void ChangeMetadataToAttributeOnItemDefinition(string projectContents, string updatedProject)
+        {
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(projectContents)),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
             Assert.Equal("m1", metadata.Name);
             Assert.Equal("v1", metadata.Value);
             Assert.False(metadata.ExpressedAsAttribute);
@@ -714,6 +961,54 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             VerifyAssertLineByLine(expected, actual);
         }
 
+        [Fact]
+        public void ChangeAttributeToMetadataOnItemDefinition()
+        {
+            var projectContents = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1 m1='v1'/>
+                        </ItemDefinitionGroup>
+                    </Project>";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(projectContents)),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(1, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].Metadata.First();
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+
+            metadata.ExpressedAsAttribute = false;
+
+            Assert.False(metadata.ExpressedAsAttribute);
+            Assert.True(project.IsDirty);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                ObjectModelHelpers.CleanupFileContents(@"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1>
+                              <m1>v1</m1>
+                            </i1>
+                        </ItemDefinitionGroup>
+                    </Project>");
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
         [Theory]
         [InlineData(@"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
@@ -769,6 +1064,49 @@ namespace Microsoft.Build.UnitTests.OM.Construction
 
             string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
                 ObjectModelHelpers.CleanupFileContents(updatedProject);
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
+        [Fact]
+        public void AddMetadataAsAttributeToItemDefinition()
+        {
+            var projectContents = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1/>
+                        </ItemDefinitionGroup>
+                    </Project>";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(projectContents)),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(0, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].AddMetadata("m1", "v1", true);
+
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+            Assert.True(project.IsDirty);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                              ObjectModelHelpers.CleanupFileContents(@"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1` />
+                        </ItemDefinitionGroup>
+                    </Project>");
             string actual = writer.ToString();
 
             VerifyAssertLineByLine(expected, actual);
@@ -845,13 +1183,65 @@ namespace Microsoft.Build.UnitTests.OM.Construction
             VerifyAssertLineByLine(expected, actual);
         }
 
+        [Fact]
+        public void AddMetadataToItemDefinitionAsAttributeAndAsElement()
+        {
+            var projectContents = @"
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
+                        <ItemDefinitionGroup>
+                            <i1/>
+                        </ItemDefinitionGroup>
+                    </Project>";
+            ProjectRootElement projectElement = ProjectRootElement.Create(XmlReader.Create(new StringReader(projectContents)),
+                ProjectCollection.GlobalProjectCollection,
+                preserveFormatting: true);
+            ProjectItemDefinitionGroupElement itemDefinitionGroup = (ProjectItemDefinitionGroupElement)projectElement.AllChildren.FirstOrDefault(c => c is ProjectItemDefinitionGroupElement);
+
+            var project = new Project(projectElement);
+
+            var itemDefinitions = Helpers.MakeList(itemDefinitionGroup.ItemDefinitions);
+
+            Assert.Equal(1, itemDefinitions.Count);
+            Assert.Equal(0, itemDefinitions[0].Metadata.Count);
+
+            var metadata = itemDefinitions[0].AddMetadata("m1", "v1", true);
+
+            Assert.Equal("m1", metadata.Name);
+            Assert.Equal("v1", metadata.Value);
+            Assert.True(metadata.ExpressedAsAttribute);
+            Assert.True(project.IsDirty);
+
+            metadata = itemDefinitions[0].AddMetadata("m2", "v2", false);
+
+            Assert.Equal("m2", metadata.Name);
+            Assert.Equal("v2", metadata.Value);
+            Assert.False(metadata.ExpressedAsAttribute);
+            Assert.True(project.IsDirty);
+
+            StringWriter writer = new StringWriter();
+            project.Save(writer);
+
+            string expected = @"<?xml version=""1.0"" encoding=""utf-16""?>" +
+                              ObjectModelHelpers.CleanupFileContents(@"
+                    <Project xmlns=`http://schemas.microsoft.com/developer/msbuild/2003`>
+                        <ItemDefinitionGroup>
+                            <i1 m1=`v1`>
+                              <m2>v2</m2>
+                            </i1>
+                        </ItemDefinitionGroup>
+                    </Project>");
+            string actual = writer.ToString();
+
+            VerifyAssertLineByLine(expected, actual);
+        }
+
         /// <summary>
         /// Helper to get a ProjectMetadataElement for a simple metadatum
         /// </summary>
         private static ProjectMetadataElement GetMetadataXml()
         {
             string content = @"
-                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                    <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003'>
                         <ItemGroup>
                             <i Include='i1'>
                                 <m Condition='c'>m1</m>

--- a/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs
+++ b/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs
@@ -1275,6 +1275,14 @@ Project(""{";
 @"<?xml version=`1.0` encoding=`utf-8`?>
   <Project xmlns = 'msbuildnamespace'>
 
+    <ItemDefinitionGroup>
+      <Compile A=`a`/>
+      <B M1=`dv1`>
+        <M2>dv2</M2>
+      </B>
+      <C/>
+    </ItemDefinitionGroup>
+
     <ItemGroup>
       <Compile Include=`Class1.cs` A=`a` />
       <Compile Include=`Class2.cs` />

--- a/src/Build/Construction/ProjectItemDefinitionElement.cs
+++ b/src/Build/Construction/ProjectItemDefinitionElement.cs
@@ -49,11 +49,31 @@ namespace Microsoft.Build.Construction
         /// </summary>
         public ProjectMetadataElement AddMetadata(string name, string unevaluatedValue)
         {
+            return AddMetadata(name, unevaluatedValue, false);
+        }
+
+        /// <summary>
+        /// Convenience method to add a piece of metadata to this item definition.
+        /// Adds after any existing metadata. Does not modify any existing metadata.
+        /// </summary>
+        /// <param name="name">The name of the metadata to add</param>
+        /// <param name="unevaluatedValue">The value of the metadata to add</param>
+        /// <param name="expressAsAttribute">If true, then the metadata will be expressed as an attribute instead of a child element, for example
+        /// &lt;Content CopyToOutputDirectory="PreserveNewest" /&gt;
+        /// </param>
+        public ProjectMetadataElement AddMetadata(string name, string unevaluatedValue, bool expressAsAttribute)
+        {
             ErrorUtilities.VerifyThrowArgumentLength(name, nameof(name));
             ErrorUtilities.VerifyThrowArgumentNull(unevaluatedValue, nameof(unevaluatedValue));
 
+            if (expressAsAttribute)
+            {
+                ProjectMetadataElement.ValidateValidMetadataAsAttributeName(name, ElementName, Location);
+            }
+
             ProjectMetadataElement metadata = ContainingProject.CreateMetadataElement(name);
             metadata.Value = unevaluatedValue;
+            metadata.ExpressedAsAttribute = expressAsAttribute;
 
             AppendChild(metadata);
 
@@ -91,5 +111,10 @@ namespace Microsoft.Build.Construction
         {
             return owner.CreateItemDefinitionElement(ItemType);
         }
+
+        /// <summary>
+        /// Do not clone attributes which can be metadata. The corresponding expressed as attribute project elements are responsible for adding their attribute
+        /// </summary>
+        protected override bool ShouldCloneXmlAttribute(XmlAttribute attribute) => !ProjectMetadataElement.AttributeNameIsValidMetadataName(attribute.LocalName);
     }
 }

--- a/src/Build/Evaluation/ProjectParser.cs
+++ b/src/Build/Evaluation/ProjectParser.cs
@@ -8,7 +8,7 @@ using Microsoft.Build.Framework;
 #if (!STANDALONEBUILD)
 using Microsoft.Internal.Performance;
 
-#if MSBUILDENABLEVSPROFILING 
+#if MSBUILDENABLEVSPROFILING
 using Microsoft.VisualStudio.Profiler;
 #endif
 #endif
@@ -130,7 +130,7 @@ namespace Microsoft.Build.Construction
                 ProjectParser parser = new ProjectParser(document, projectRootElement);
                 parser.Parse();
             }
-#if MSBUILDENABLEVSPROFILING 
+#if MSBUILDENABLEVSPROFILING
             }
             finally
             {
@@ -406,7 +406,7 @@ namespace Microsoft.Build.Construction
         }
 
         /// <summary>
-        /// Parse a ProjectMetadataElement 
+        /// Parse a ProjectMetadataElement
         /// </summary>
         private ProjectMetadataElement ParseProjectMetadataElement(XmlElementWithLocation element, ProjectElementContainer parent)
         {
@@ -636,7 +636,7 @@ namespace Microsoft.Build.Construction
 
                     case XMakeElements.onError:
                         // Previous OM accidentally didn't verify ExecuteTargets on parse,
-                        // but we do, as it makes no sense 
+                        // but we do, as it makes no sense
                         ProjectXmlUtilities.VerifyThrowProjectAttributes(childElement, ValidAttributesOnOnError);
                         ProjectXmlUtilities.VerifyThrowProjectRequiredAttribute(childElement, XMakeAttributes.executeTargets);
                         ProjectXmlUtilities.VerifyThrowProjectNoChildElements(childElement);
@@ -748,13 +748,33 @@ namespace Microsoft.Build.Construction
         /// </summary>
         private ProjectItemDefinitionElement ParseProjectItemDefinitionXml(XmlElementWithLocation element, ProjectItemDefinitionGroupElement parent)
         {
-            ProjectXmlUtilities.VerifyThrowProjectAttributes(element, ValidAttributesOnlyConditionAndLabel);
-
             // Orcas inadvertently did not check for reserved item types (like "Choose") in item definitions,
             // as we do for item types in item groups. So we do not have a check here.
-            // Although we could perhaps add one, as such item definitions couldn't be used 
+            // Although we could perhaps add one, as such item definitions couldn't be used
             // since no items can have the reserved itemType.
             ProjectItemDefinitionElement itemDefinition = new ProjectItemDefinitionElement(element, parent, _project);
+
+            foreach (XmlAttributeWithLocation attribute in element.Attributes)
+            {
+                CheckMetadataAsAttributeName(attribute.Name, out bool isKnownAttribute, out bool isValidMetadataNameInAttribute);
+
+                if (!isKnownAttribute && !isValidMetadataNameInAttribute)
+                {
+                    ProjectXmlUtilities.ThrowProjectInvalidAttribute(attribute);
+                }
+                else if (isValidMetadataNameInAttribute)
+                {
+                    ProjectMetadataElement metadatum = _project.CreateMetadataElement(attribute.Name, attribute.Value);
+                    metadatum.ExpressedAsAttribute = true;
+                    metadatum.Parent = itemDefinition;
+
+                    itemDefinition.AppendParentedChildNoChecks(metadatum);
+                }
+                else if (!ValidAttributesOnlyConditionAndLabel.Contains(attribute.Name))
+                {
+                    ProjectXmlUtilities.ThrowProjectInvalidAttribute(attribute);
+                }
+            }
 
             foreach (XmlElementWithLocation childElement in ProjectXmlUtilities.GetVerifyThrowProjectChildElements(element))
             {

